capsules/src/ble_advertising_driver.rs

```diff
@@ -445,120 +218,59 @@ impl App {
-                            for (dst, src) in data.as_mut()[idx + 2..end]
-                                .iter_mut()
-                                .zip(slice.as_ref()[0..slice.len()].iter())
-                            {
-                                *dst = *src;
-                            }
-                            ReturnCode::SUCCESS
-                        })
-                        .unwrap_or_else(|| ReturnCode::EINVAL);
-
-                    // If the operation was successful => update idx
-                    if result == ReturnCode::SUCCESS {
-                        self.idx = end;
-                    }
-                    result
-                } else {
-                    ReturnCode::ESIZE
-                }
-            })
-            .unwrap_or_else(|| ReturnCode::EINVAL)
+        self.address = [
+            0xf0,
+            (appid.idx() & 0xff) as u8,
+            ((appid.idx() << 8) & 0xff) as u8,
+            ((appid.idx() << 16) & 0xff) as u8,
+            ((appid.idx() << 24) & 0xff) as u8,
+            0xf0,
+        ];
+        ReturnCode::SUCCESS
     }
 
     fn send_advertisement<'a, B, A>(&self, ble: &BLE<'a, B, A>, channel: RadioChannel) -> ReturnCode
     where
         B: ble_advertising::BleAdvertisementDriver + ble_advertising::BleConfig + 'a,
         A: kernel::hil::time::Alarm + 'a,
     {
-        self.advertisement_buf
+        self.adv_data
             .as_ref()
-            .map(|slice| {
+            .map(|adv_data| {
                 ble.kernel_tx
                     .take()
-                    .map(|data| {
-                        for (out, inp) in data.as_mut()[PACKET_HDR_PDU..PACKET_LENGTH]
-                            .iter_mut()
-                            .zip(slice.as_ref()[PACKET_HDR_PDU..PACKET_LENGTH].iter())
+                    .map(|kernel_tx| {
+                        let adv_data_len =
+                            cmp::min(kernel_tx.len() - PACKET_ADDR_LEN - 2, adv_data.len());
+                        let adv_data_corrected = &adv_data.as_ref()[..adv_data_len];
+                        let payload_len = adv_data_corrected.len() + PACKET_ADDR_LEN;
                         {
-                            *out = *inp;
+                            let (header, payload) = kernel_tx.split_at_mut(2);
+                            header[0] = self.pdu_type;
+                            match self.pdu_type {
+                                ADV_IND | ADV_NONCONN_IND | ADV_SCAN_IND => {
+                                    // Set TxAdd because AdvA field is going to be a "random"
+                                    // address
+                                    header[0] |= 1 << ADV_HEADER_TXADD_OFFSET;
+                                }
+                                _ => {}
+                            }
+                            // The LENGTH field is 6-bits wide, so make sure to truncate it
+                            header[1] = (payload_len & 0x7f) as u8;
```

Comment:
Should this be `0x3f`?

Reply:
Yes, I think so. 

@alevy can you confirm?

Reply:
Damnit... yes :)